### ISSUE TITLE
Support enums for more accurate static types

### ISSUE THREAD

**Issue by Pauan** [rustwasm/wasm-bindgen#2088]
### Motivation

Because JavaScript is dynamically typed, it often mixes and matches types in a way which violates static type checking.

For example, consider this TypeScript function:

```ts
function foo(): Date | number {
    if (someCondition()) {
        return new Date();
    } else {
        return 5;
    }
}
```

This function cannot be given a proper type in wasm-bindgen, so instead it must return `JsValue`. But this is quite unfortunate.

### Proposed Solution

We already support C-style enums, and enums with string values. We can extend that to support ADTs where each variant contains a single JS value:

```rust
#[wasm_bindgen]
pub enum Foo {
    Date(js_sys::Date),
    Number(js_sys::Number),
}
```

When converting from a `JsValue` into a `Foo`, it will do runtime type checking (using `JsCast`) and choose the first variant which matches.

Now we can say that the `foo` function returns `Foo`, and static typing is restored.

This could also potentially be used to improve our web-sys bindings as well.

### Alternatives

We could create some sort of wasm-bindgen specific solution, but I think reusing Rust enums makes a lot of sense, since it gives us pattern matching and exhaustiveness checking.

### Additional Context

I'm willing to make a PR for this.

**Comment by workingjubilee** [comment]
I would like to voice additional support for both this and also add that I would like to see additionally a way to write this specifically for the primitive types of JS (essentially, as someone who touched JS before I touched Rust, I want to be able to turn to Rust and say "use typeof here"), making

```js
switch (typeof unknownJsValue) {
}
```
equivalent to
```rust
let unknown_jstype = unknown_jsvalue./* new typegetter method on JsCast */

if let Some(x) = unknown_jstype {
   match jstype {
     String(str_value) => do_something(str_value),
     /* et cetera */
   }
}
```

#### A Slight Problem
This variation would not intelligently execute on the differentiation between a `"string literal"` and a `String("constructor")` which is an object, but as it is unlikely people will use String constructors in JS in that way, I think including a warning about this technical-but-unlikely fallibility when using this is sufficient.

**Comment by Pauan** [comment]
@workingjubilee That's already possible today: you can use [`as_string()`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/struct.JsValue.html#method.as_string) to convert into a string (and similarly for other primitive types).

Or alternatively you can use [`JsCast`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/trait.JsCast.html) to check for [`JsString`](https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.JsString.html):

```rust
match foo.dyn_into::<JsString>() {
    Ok(string) => { ... },
    Err(foo) => { ... },
}
```

The point of this feature request is to make it a lot easier to use, and also to get correct TypeScript types.

**Comment by workingjubilee** [comment]
That will do for one type, and will not do if I want to be able to write a match over the _entire_ set of JS primitive types, which was the point of what I was saying.

**Comment by Pauan** [comment]
@workingjubilee That seems like a completely different request than this issue.

You can do this to access `typeof`:

```rust
#[wasm_bindgen(inline_js = "export function typeof_(x) { return typeof x; }")]
extern "C" {
    fn typeof_(x: &JsValue) -> String;
}
```

Now you can do this:

```rust
match typeof_(&foo).as_str() {
    "undefined" => ...
    "object" if foo.is_null() => ...
    "object" => ...
    "boolean" => ...
    "number" => ...
    "bigint" => ...
    "string" => ...
    "symbol" => ...
    "function" => ...
}
```

However that won't do any type conversions, which is why we provide methods like `as_string`, because they do the type check *and* the type conversion.

So if you wanted it to do the type conversion, you would do something like this:

```rust
enum Typeof {
    Undefined,
    Null,
    Boolean(bool),
    Number(f64),
    String(String),
    Function(Function),
    Object(JsValue),
}

impl From<JsValue> for Typeof {
    fn from(value: JsValue) -> Self {
        if value.is_undefined() {
            return Self::Undefined;
        }

        if value.is_null() {
            return Self::Null;
        }

        if let Some(value) = value.as_bool() {
            return Self::Boolean(value);
        }

        if let Some(value) = value.as_f64() {
            return Self::Number(value);
        }

        if let Some(value) = value.as_string() {
            return Self::String(value);
        }

        match value.dyn_into::<Function>() {
            Ok(value) => Self::Function(value),
            Err(value) => Self::Object(value),
        }
    }
}
```

Now you can just do this:

```rust
Typeof::from(foo)
```

But if you explained what you're trying to accomplish, there is probably a better way to do it.

**Comment by workingjubilee** [comment]
Recursive analysis of arbitrary JS code! :smile:

Interesting, though, and noted. I somehow managed to read most of the API docs without encountering the import function example in spite of it being relatively prominently located. Maybe I should get my eyes checked!

**Comment by alexcrichton** [comment]
This sounds like a great idea to me!

I think we'll want to be pretty restrictive about enums-with-payloads, basically require them to all have one variant which is effectively a `JsValue`, but otherwise I think this'd be an awesome feature to have.

**Comment by Pauan** [comment]
@alexcrichton Yes, that was my original idea (and implementation): the values would all have to implement `JsCast`. But I've been working on a system which will generalize things a lot better (right now the conversion between JS and Rust is kind of a mess).

**Comment by shortenda** [comment]
Just curious, is there a reason not to expose non-c-style enums as typescript discriminated unions? (https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions)

**Comment by Pauan** [comment]
@shortenda There isn't a standardized syntax for it:

* Should the discriminant be called `kind` or `type`? Or something else? Should it be customizable?

* Should the discriminant use numbers or strings?

* How should tuple variants be handled?

* How would it distinguish between object enums and non-object enums? Should there be an attribute for it?

Somebody would have to figure out an answer to these questions first.
